compiler/rustc_parse/src/parser/generics.rs

```diff
@@ -279,7 +279,27 @@ impl<'a> Parser<'a> {
         let span_lo = self.token.span;
         let (params, span) = if self.eat_lt() {
             let params = self.parse_generic_params()?;
-            self.expect_gt()?;
+            if let Err(mut err) = self.expect_gt() {
+                // Try to recover a `:` into a `::`
+                // Attempt to find places where a missing `>` might belong.
```

Comment:
can you please hide this behind a function call. I don't think it's good to clutter core functions like `parse_generics` with diagnostics-related code.

Also can you maybe include the example from the `tests/ui/generics/unclosed-generics-in-impl-def.rs` test in this comment?